Configuration Generation

Timidger opened the pull request:
If ~/.mackup.cfg does not exist, Mackup asks the user if he wants to make one. It then prompts them to answer some questions so that it can be generated. Also generates custom apps in the ~/.mackup folder if the user wants to.

lra commented:
After giving it more thoughts, I made up my mind of having mackup to syncing anything by default. See https://github.com/lra/mackup/issues/614

But this is still super useful, except that it should be triggered by the user.

Can you make it an explicit action (e.g. `mackup configure`) instead of checking if the config file exist?

Timidger commented:
Being explicit here is the better option, especially since a user might not want to answer a bunch of questions after downloading some software for the first time.  I'll get to work switching this over to be an explicit command.

Timidger commented:
I switched it over to a command (`mackup configure`), and also updated the storage options to include the new ones since the patch was first written.

lra commented on mackup/config.py:
```diff
@@ -25,6 +25,51 @@
+    if not blacklist:
+        blacklist = []
+    # Make sure the storage_type is correct
+    assert storage_type in (ENGINE_DROPBOX, ENGINE_FS,
```
we have a lot more now

lra commented on mackup/config.py:
```diff
@@ -25,6 +25,51 @@
+    assert storage_type in (ENGINE_DROPBOX, ENGINE_FS,
+                            ENGINE_GDRIVE)
+    # If the path is specified, make sure it exists
+    if path:
```
only allow this if the ENGINE_FS is used.

lra commented on mackup/config.py:
```diff
@@ -25,6 +25,51 @@
+        assert os.path.exists(path), (
+            "The path {} does not exist!".format(path))
+    # Write the configuration file in the home directory
+    configuration = ["[storage]",
```
Can you use Python's ConfigParser to create this file content? Way easier and solid than trying to build it with strings.

lra commented on mackup/config.py:
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
+    def make_config_file():
+        storage_type = choose("What method would you like to use to backup "
+                              "your configuration files?",
+                              ["dropbox", "google_drive", "file_system"])[0]
```
we have more

lra commented on mackup/utils.py:
```diff
@@ -9,6 +9,50 @@
+    default_str = ("(default: {})".format(default) if default in options
+                   else "")
+    while True:
+        answer = raw_input("{} {} <{}>\n".format(question, default_str,
```
if default_str is empty, you'll have 2 spaces, which look bad.

lra commented on mackup/config.py:
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
+        whitelist = ""
+        if not confirm("Would you like to sync all applications or "
+                              "specify from a list of supported software?"):
+            whitelist = choose("Choose any of the follwing "
```
following

lra commented on mackup/config.py:
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
+        blacklist = ""
+        if confirm("Would you like to specify any applications you would "
+                        "NOT like to sync?"):
+            blacklist = choose("Choose any of the follwing "
```
following

lra commented on mackup/config.py:
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
+        custom_apps = {} # {app_name: [config files]}
+        if confirm("Now, would you like to add any applications not "
+                         "currently supported by Mackup?"):
+            print "Enter nothing to indicate your done"
```
that you are done

lra commented on mackup/config.py:
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
+                if not name:
+                    break
+                config_files = raw_input("Please enter the names of all the "
+                "configuration files, delemenated by a space").split()
```
delimited

lra commented on mackup/config.py:
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
+                name = raw_input("Application name: ")
+                if not name:
+                    break
+                config_files = raw_input("Please enter the names of all the "
```
This is not gonna work, as paths can have spaces in them.

lra commented on mackup/config.py:
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
+                custom_apps.update({name: config_files})
+        with open(os.path.join(os.path.expanduser("~"),
+                   MACKUP_CONFIG_FILE), "w") as config_file:
+            config_file.write("[storage]\n")
```
Same here, you should use Python's ConfigParser

lra commented on mackup/utils.py:
```diff
@@ -407,3 +451,122 @@ def can_file_be_synced_on_current_platform(path):
+
+def get_storage_config():
+    """
+    Asks the user to enter storage information written that will later be
```
double written

lra commented on mackup/utils.py:
```diff
@@ -407,3 +451,122 @@ def can_file_be_synced_on_current_platform(path):
+                print "The path {} does not exist!".format(exists)
+
+    while True:
+        directory = (raw_input("What would directory name would you like to "
```
Which instead of What would

lra commented on mackup/utils.py:
```diff
@@ -407,3 +451,122 @@ def can_file_be_synced_on_current_platform(path):
+        assert os.path.exists(path), (
+            "The path {} does not exist!".format(path))
+    # Write the configuration file in the home directory
+    configuration = ["[storage]",
```
Same, use ConfigParser

Timidger commented:
Switched over the configuration file generation from the hacky string-based one to ConfigParser's built-in mechanism for it. I had to build it backwards though (the docs says it writes it backwards for some reason). Tested it myself as well, it generates perfectly.

Timidger commented:
Updated the files. I deleted that config generator function in `utils.py`, since I moved it to `config.py` (all of its helper functions are still in `utils.py` though).

paour commented:
Too bad this PR is languishing, it fixes a real pain point in mackup…

sethherr commented:
:+1: I didn't know .mackup.cfg existed until looking for .gitignore_global, having documentation of it and a tool to use it (my first .mackup.cfg attempt failed because I didn't include headers) would be great.

lra commented:
@sethherr PR welcome ;)
